add proposal for Substratus managed infra (#171)
substratusai/substratus: docs/proposals/operator-managed-infra.md

* Utilize a single service account that has enough permissions to do everything needed within Substratus
* Simplify Substratus to use a single K8s SA per namespace. Simplicity is preferred here over the minor security benefits you get
  for using a different SA for each kind of Substratus resource.
* Provide Substratus controller the permissions to manage registry, bucket and set IAM policy on the GCP SA.
  The following permissions will be required by the substratus Service Account on the project itself:
  ```
  roles/storage.admin, roles/artifactregistry.repoAdmin, roles/iam.serviceAccountAdmin
  ```

### Implementation details
* Object storage: Create bucket if the configured bucket does not exist
* Image Registry: Create registry if the configured registry does not exist
* New sub reconciler for namespace: Create K8s Service Accounts automatically in each namespace where Substratus is used and
  annotate the K8s Service Account with the correct GCP Service Account. Ensure that following
  gets called when a new Service Account is provisioned in a new namespace:
  ```
  gcloud iam service-accounts add-iam-policy-binding substratus@my-project.iam.gserviceaccount.com \

>    --role roles/iam. workloadIdentityUser \

Reviewer: ```suggestion
   --role roles/iam.workloadIdentityUser \
```

   --member "serviceAccount:myproject.svc.id.goog[new-namespace/substratus]"
  ```

>   Make sure this only happens whenever a new Substratus gets created in a new namespace.

Reviewer: ```suggestion
  Make sure this only happens whenever a new Substratus resource is created in a new namespace.
```

  The controller should be smart enough to not have to create KSAs in namespaces that

>   do not have any substratus resource.

Reviewer: ```suggestion
  do not have any substratus resources.
```

How should Substratus controller handle infra management?
* Terraform

> * Native K8s controller that uses Golang SDK for AWS/GCP/..

Reviewer: strong preference for just using the go sdk within the controller.

The Native K8s controller would better fit within the existing Substratus code base.
It would also allow us to verify if a resource is already there and if not just create
it. Terraform would still be used for creation of GKE/EKS cluster to provide a reference
example for easy install.

### Open Questions

1. Where should the code to create Image Registry and Bucket live? Do we need a new
   resource or can we reconcile based purely on configuration?

> 2. Should Substratus install and manage Karpenter on AWS?

Reviewer: I believe we landed on no for this and similar cluster-level DaemonSets (e.g., nvidia drivers) though cluster configurations like this would be good to verify through the controller. If incompatible configurations are found, the controller should likely emit warning level logs.

3. Should we change the nodeSelectors based on whether Karpenter is or isn't installed?

> 4. Should we install and manage the Nvidia DaemonSet?

Reviewer: I believe we landed on no for this and similar cluster-level DaemonSets (e.g., nvidia drivers) though cluster configurations like this would be good to verify through the controller. If incompatible configurations or non-existent dependencies are found, the controller should likely emit warning level logs.

I'm still not 100% convinced that SCI installation shouldn't be managed by the core controller. I think that's worth revisiting as we get further.

# Operator Managed Infra and multi namespace
Design doc: Not a feature yet

Instead of managing all the infrastructure through terraform, the
operator itself should be responsible for managing the Workload Identity
bindings between the K8s Service Account and the Cloud Service Account/IAM role.

Doing this, will make it possible to use Substratus resources in multiple
namespaces without having to use another tool to configure Cloud identity/role mappings.

Out of scope:

> * Managing the bucket, image registry or other any other resources

Reviewer: ```suggestion
* Managing the cluster configuration, node pools, dependent workpads/daemonsets, the bucket, image registry, the `sci-${provider}` principal, and its permissions.
```

## How?

<img src="../diagrams/identity-global.excalidraw.png"></img>

### Installation
**Installation Flow** will be responsible for the following:
* Create a K8s cluster with nodes/GPUs, this step is optional if end-user already has a K8s cluster (eksctl or gcloud)
* Create Bucket, this step is optional if end-user already has a bucket
* Create Image Registry, this step is optional if end-user already has an image registry
* REQUIRED: Create Google Service Account or AWS IAM Role that has resource level permissions to:
  * read and write a specific bucket
  * pull and push to a specific Image Registry (GAR/ECR)
  * (GCP) set IAM policy on the service Account itself to be able to manage workload identity bindings to KSAs. This can be done by assigning the service account to be a ServiceAccountAdmin to itself
  * (AWS) UpdateAssumeRolePolicy on the AWS Role used by Substratus (required for multi namespace support)
* REQUIRED: Apply Substratus manifests and configure them to use correct bucket and image registry

### Controller
Controller will be responsible for the following:

> * Creating the K8s ServiceAccount (already done today) and related binding by calling `enforceServiceAccount` whenever a Substratus resource gets reconciled

Reviewer: To make the call provider agnostic, I was leaning toward `UpdateIamPrincipal` as the GRPC service we'd implement. The term principal seems to be common terminology across all providers. `Update` is in line with other verbs I've seen in GRPC-land.

```suggestion
* Creating the K8s ServiceAccount (already done today) and related binding by calling `UpdateIamPrincipal` whenever a Substratus resource gets reconciled
```

Variations:

* `UpdateIamPrincipalPolicy`
* `UpdateIdentityPolicy`

Author: I hadn't thought of this being a generic call but that makes total sense.

Author: I will rephrase this since we should re-use the existing enforceServiceAccount function but in that function i should call the generic `UpdateIamPrincipal` from `sci-${CLOUD}`

> * (GCP only) example set annotation for Google Service Account (already done today) AND call CSI to update IAM policy on the Service Account so it can use the Google Service Account. For example:

Reviewer: ```suggestion
* (GCP only) example set annotation for Google Service Account (already done today) AND call SCI to update IAM policy on the Service Account so it can use the Google Service Account. For example:
```

  ```
  gcloud iam service-accounts add-iam-policy-binding substratus@my-project.iam.gserviceaccount.com \
   --role roles/iam.workloadIdentityUser \
   --member "serviceAccount:myproject.svc.id.goog[new-namespace/substratus]"
  ```

> * (AWS Only) Annotate the service account with ARN of IAM role AND (awsmanager) use CSI to call UpdateAssumeRolePolicy

Reviewer: This is new as of this afternoon but `sci-${provider}` is better because it's more specific than our current`${provider}-manager`.

```suggestion
* (AWS Only) Annotate the service account with ARN of IAM role AND (sci-aws) use CSI to call UpdateAssumeRolePolicy
```

* Create a K8s cluster with nodes/GPUs, this step is optional if end-user already has a K8s cluster (eksctl or gcloud)
* Create Bucket, this step is optional if end-user already has a bucket
* Create Image Registry, this step is optional if end-user already has an image registry
* REQUIRED: Create Google Service Account or AWS IAM Role that has resource level permissions to:
  * read and write a specific bucket
  * pull and push to a specific Image Registry (GAR/ECR)
  * (GCP) set IAM policy on the service Account itself to be able to manage workload identity bindings to KSAs. This can be done by assigning the service account to be a ServiceAccountAdmin to itself
  * (AWS) UpdateAssumeRolePolicy on the AWS Role used by Substratus (required for multi namespace support)
* REQUIRED: Apply Substratus manifests and configure them to use correct bucket and image registry

### Controller
Controller will be responsible for the following:
* Creating the K8s ServiceAccount (already done today in `enforceServiceAccount`), in addition should now also call `sci-${CLOUD}:BindKSAToIAMPrincipal` to ensure the new KSA has access to the AWS IAM Role/GCP Service Account

> * (GCP only) example set annotation for Google Service Account (already done today) AND call `sci-gcp:BindKSAToIAMPrincipal` to update IAM policy on the Service Account so it can use the Google Service Account. For example:

Reviewer: I think the call would just be `sci.BindKSAToIAMPrincipal` across all providers.

Should we separate out the GRPC services for adding and removing bindings? That seems implied in the decision to use `BindKSAToIAMPrincipal`. I was using `UpdateIamPrincipal` because I imagined an `operation: string` argument in the function signature with valid values of either `"add"` or `"remove"` so we'd be able to reuse the same service for both use cases. Let's either decide to:

1. consolidate into a single GRPC service or
2. decide on both add and removal service names (they're both needed) and add a detail on how that would look, what would trigger it, etc.

Author: I think we should never remove a K8s SA from an IAM principal. So vote for leaving that out of the proposal until we later see a need for it? When should we remove a K8s SA binding from IAM principal? I think it's better to leave those intact until the IAM principal itself is removed.

Reviewer: > When should we remove a K8s SA binding from IAM principal?

If a namespace is deleted

Author: I'm ore concerned about the risk where an end-user has 2 clusters and in 1 cluster we remove the binding which causes his other cluster to also lose access unexpectedly. Hence I think we should just leave them in-place. Does AWS do cluster level bindings? If so that would be a good argument for implementing both add and remove. In GCP remove simply wouldn't do anything.

Reviewer 2: Prefer to table cleanup for another day. Lets capture the concerns in an issue.

Reviewer: Captured here: https://github.com/substratusai/substratus/issues/179

  ```
  gcloud iam service-accounts add-iam-policy-binding substratus@my-project.iam.gserviceaccount.com \
   --role roles/iam.workloadIdentityUser \
   --member "serviceAccount:myproject.svc.id.goog[new-namespace/substratus]"
  ```
* (AWS Only) Annotate the Kubernetes service account with ARN of IAM role AND (sci-aws) use `sci-aws:BindKSAToIAMPrincipal` to call `UpdateAssumeRolePolicy` to update the role's trust policy.
* Any API calls made to clouds should go through a cloud manager e.g. `gcpmanager` or `awsmanager`

### Substratus Cloud Interface
The SCI will have a generic function `BindKSAToIAMPrincipal` that takes the following input:
```
namespace: The K8s namespace

> serviceAccount: The name of K8s Service Account

Reviewer: ```suggestion
kubernetesServiceAccount: The name of K8s Service Account
```

> IAMPrinciple: The GCP service account or IAM role

Reviewer: ```suggestion
IAMPrincipal: The GCP service account or IAM role
```

> Params Map[string]string: Any additional parameters that might be required by a specific cloud provider

Reviewer 2: Prefer to omit Params until we find the need

Reviewer 2: Defer to convo here: https://github.com/substratusai/substratus/pull/177

Reviewer 2: OK with leaving this level of detail out of proposal, or just stating that these things are "first thoughts" in order to get proposal closed.

Author: I think when possible we should discuss it in the proposal instead of having to significantly change the code PR later. I will add a note saying the actual implementation will likely differ

Author: agreed with Nick that some of these should be provided as part of configMap or automatically detected at runtime instead of using params. This is such that the controller doesn't have to have different call for each cloud provider

Reviewer: My take: proposals are good but probably shouldn't hammer out fine grain details. Proposals should paint the high-level picture with a rough outline of the implementation plan. Things like naming should often be deferred

Author: I think the risk with that is renaming in a PR with code sometimes requires full retesting and changing a lot of code. So I rather hash out as many details as possible. The interfaces and names of those interfaces should be discussed in the proposal when possible but not be required.

  * pull and push to a specific Image Registry (GAR/ECR)
  * (GCP) set IAM policy on the service Account itself to be able to manage workload identity bindings to KSAs. This can be done by assigning the service account to be a ServiceAccountAdmin to itself
  * (AWS) UpdateAssumeRolePolicy on the AWS Role used by Substratus (required for multi namespace support)
* REQUIRED: Apply Subsubstratus manifests and configure them to use correct bucket and image registry

### Controller
Controller will be responsible for the following:
* Creating the K8s ServiceAccount and related binding by calling `enforceServiceAccount` whenever a Substratus resource gets reconciled
* (GCP only) example set annotation for Google Service Account AND update IAM policy on the Service Account so it can use the Google Service Account. For example:
  ```
  gcloud iam service-accounts add-iam-policy-binding substratus@my-project.iam.gserviceaccount.com \
   --role roles/iam.workloadIdentityUser \
   --member "serviceAccount:myproject.svc.id.goog[new-namespace/substratus]"
  ```
* (AWS Only) Annotate the service account with ARN of IAM role AND (awsmanager) call UpdateAssumeRolePolicy

> * The annotations to service accounts happen inside the controller itself

Reviewer 2: Already stated

* Any API calls made to clouds should go through a cloud manager e.g. `gcpmanager` or `awsmanager`

## User Impact / Docs
TODO: Update this with actual proposed install steps for GCP

1. (Optional) Create your GKE cluster `gcloud container clusters create` and create nodepools

2. Create service account and assign permissions needed

   ```
   gcloud iam service-accounts create substratus
   gcloud iam service-accounts add-iam-policy-binding substratus@my-project.iam.gserviceaccount.com \
    --role roles/iam.serviceAccountAdmin \

>     --member "serviceAccount:myproject.svc.id.goog[substratus/substratus]"

Reviewer 2: I think `--member` is wrong here, should likely be `--member "serviceAccount:substratus@my-project.iam.gserviceaccount.com"`

Reviewer 2: I am guessing this was to give this SA access to modify its own bindings?

Reviewer 2: PS: We will also need workload identity binding for substratus to start using the SA:

```bash
gcloud iam service-accounts add-iam-policy-binding GSA_NAME@GSA_PROJECT.iam.gserviceaccount.com \
    --role roles/iam.workloadIdentityUser \
    --member "serviceAccount:PROJECT_ID.svc.id.goog[NAMESPACE/KSA_NAME]"
```

Author: this was the part with a todo that I didn't get to update yet :P I added this at the top "TODO: Update this with actual proposed install steps for GCP"

in hindsight I should have just removed it before asking you all to review

>    gcloud projects add-iam-policy-binding my-project \

Reviewer 2: Lets avoid project level bindings where we can: https://cloud.google.com/artifact-registry/docs/access-control#grant-repo

Author: yep latest version doesn't have this, you weren't supposed to review this part, sorry about that

Design doc: Not a feature yet

Instead of managing all the infrastructure through terraform, the
operator itself should be responsible for managing the Workload Identity
bindings between the K8s Service Account and the Cloud Service Account/IAM role.

Doing this, will make it possible to use Substratus resources in multiple
namespaces without having to use another tool to configure Cloud identity/role mappings.

Out of scope:
* Managing the bucket, image registry or other any other resources

## Why?
* Simplified install: Allow install to use `gcloud/aws` or `eksctl` to create initial K8s cluster + nodepools
* Make install onto existing K8s cluster straight forward and remove
  complexity of using Terraform

> * Improved UX, all an end-user needs to do is install our manifests and ensure GCP service account has enough permissions

Reviewer: > ensure GCP service account has enough permissions

I think fleshing out what's meant by this statement will answer what we want here: https://github.com/substratusai/substratus/pull/170#discussion_r1287980291

### Installation
**Installation Flow** will be responsible for the following:
* Create a K8s cluster with nodes/GPUs, this step is optional if end-user already has a K8s cluster (eksctl or gcloud)
* Create Bucket, this step is optional if end-user already has a bucket
* Create Image Registry, this step is optional if end-user already has an image registry
* REQUIRED: Create Google Service Account or AWS IAM Role that has resource level permissions to:
  * read and write a specific bucket
  * pull and push to a specific Image Registry (GAR/ECR)
  * (GCP) set IAM policy on the service Account itself to be able to manage workload identity bindings to KSAs. This can be done by assigning the service account to be a ServiceAccountAdmin to itself
  * (AWS) UpdateAssumeRolePolicy on the AWS Role used by Substratus (required for multi namespace support)
* REQUIRED: Apply Subsubstratus manifests and configure them to use correct bucket and image registry

### Controller
Controller will be responsible for the following:
* Creating the K8s ServiceAccount and related binding by calling `enforceServiceAccount` whenever a Substratus resource gets reconciled

> * (GCP only) example set annotation for Google Service Account AND update IAM policy ton the Service Account so it can use the Google Service Account. For example:

Reviewer: ```suggestion
* (GCP only) example set annotation for Google Service Account AND update IAM policy on the Service Account so it can use the Google Service Account. For example:
```

  gets called when a new Service Account is provisioned in a new namespace:
  ```
  gcloud iam service-accounts add-iam-policy-binding substratus@my-project.iam.gserviceaccount.com \

>    --role roles/iam. workloadIdentityUser \

Reviewer: ```suggestion
--role roles/iam.workloadIdentityUser \
```

    --role "roles/storage.admin" --role "roles/artifactregistry.repoAdmin"
   ```

>    The role `iam.serviceAccountAdmin`` is needed to be able to use workload identity across multiple

Reviewer: ```suggestion
The role `iam.serviceAccountAdmin` is needed to be able to use workload identity across multiple
```
